src/NHibernate/Impl/SessionFactoryImpl.cs

```diff
@@ -247,7 +248,7 @@ public SessionFactoryImpl(Configuration cfg, IMapping mapping, Settings settings
 					if (cache != null)
 					{
 						caches.Add(cacheRegion, cache);
-						allCacheRegions.Add(cache.RegionName, cache.Cache);
+						((IDictionary<string, ICache>)allCacheRegions).Add(cache.RegionName, cache.Cache);
```

Comment:
The documentation, however, states:
All public and protected members of ConcurrentDictionary<TKey, TValue> are thread-safe and may be used concurrently from multiple threads. However, members accessed through one of the interfaces the ConcurrentDictionary<TKey, TValue> implements, including extension methods, are not guaranteed to be thread safe and may need to be synchronized by the caller. 

We shouldn't rely on behavior the documentation doesn't promise.